enhancements/builds/volume-mounted-resources.md

```diff
@@ -0,0 +1,298 @@
+Similarly a `VolumeMount[]` field will be added, but without the MountPropagation and SubPath fields.
+MountPropagation and SubPath can be considered for support in the future.
+
+These fields will be wired, via the build controller, directly to the build pod that is constructed.
+In addition all mounts into the pod will be done at a path of our choosing, not the VolumeMount path specified, to ensure the user cannot overwrite critical function inside the build pod and use it as an escalation pathway.
+Example - all volumes are mounted under `/var/run/openshift.io/volumes` in the build pod containers.
+
+The logic that invokes buildah will then pass the mounted directories as transient volume mount arguments.
+The mount path provided to buildah will be determined from the VolumeMount specification.
+Arguments such as "read only" will also be inferred.
+
+Builds today have default mounts that are provided via other means:
+
+1. Entitlement data from the node, mounted at `/run/secrets/etc-pki-entitlement`, `/run/secrets/redhat.repo`, and `run/secrets/rhsm` inside the build pod.
+2. The cluster trust bundle - this is optionally mounted at `/etc/pki/ca-trust` via the BuildConfig spec's `mountTrustedCA` option.
+
+If a build volume's mount path conflicts with any of the paths specified above, the build volume mount will override the default.
+
+Proposed api/structs:
+Note:  DockerBuildStrategy will be updated in the same way.
+
+```go
+// SourceBuildStrategy defines input parameters specific to an Source build.
+type SourceBuildStrategy struct {
+  // From is reference to an DockerImage, ImageStream, ImageStreamTag, or ImageStreamImage from which
+  // the docker image should be pulled
+  From kapi.ObjectReference
+
+  // PullSecret is the name of a Secret that would be used for setting up
+  // the authentication for pulling the Docker images from the private Docker
+  // registries
+  PullSecret *kapi.LocalObjectReference
+
+  // Env contains additional environment variables you want to pass into a builder container.
+  Env []kapi.EnvVar
+
+  // Scripts is the location of Source scripts
+  Scripts string
+
+  // Incremental flag forces the Source build to do incremental builds if true.
+  Incremental *bool
+
+  // ForcePull describes if the builder should pull the images from registry prior to building.
+  ForcePull bool
+
+  // Volumes is a list of volumes that can be mounted by the build
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes
+  // Only Secret and ConfigMap type volumes are supported currently.
+  Volumes []BuildVolume
+
+  // VolumeMounts are volumes to mount into the build
+  VolumeMounts []VolumeMount
+}
+```
+
+```go
+// BuildVolume describes a volume that is made available to build pods, such that it can be mounted into the build execution environment.
+// Only a subset of Kubernetes Volume sources are supported by builds.
+type BuildVolume struct {
```

Comment:
doing this, instead of reusing the k8s volume type and adding validation/constraints that control which volumetypes are allowed may be a sticking point for people like @deads2k or @smarterclayton 

personally i'm ok with it (though it does carry a slightly higher maintenance burden for you), but i'd like to hear from them just in case they are fundamentally opposed.

Reply:
I decided to go this route because the real meat of the k8s Volume API is in the `VolumeSource` object, which to my understanding is a bag of pointers. We get the added maintenance burden of validating one and only one source type is set, which we don't _have_ to do on admission (though this is probably a good idea).

Reply:
On the flip side, we benefit by gating volume types in the API - invalid volume types fail fast.

Reply:
"validating one and only one source type is set" can be done purely in OpenAPI, no need for admission.

Reply:
> doing this, instead of reusing the k8s volume type and adding validation/constraints that control which volumetypes are allowed may be a sticking point for people like @deads2k or @smarterclayton

Having the local help and the constrained choice and avoiding automatic additions are all good properties you get with this approach.